[lldb] Add support for synthetic LineEntry objects without valid address ranges

medismailben opened the pull request:
Scripted frames that materialize Python functions are PC-less by design,
meaning they don't have valid address ranges. Previously, LineEntry::IsValid()
required both a valid address range and a line number, preventing scripted
frames from creating valid line entries for these synthetic stack frames.

Relaxing this requirement is necessary since `SBSymbolContext::SetLineEntry`
will first check if the LineEntry is valid and discard it otherwise.

This change introduces an `synthetic` flag that gets set when LineEntry
objects are created or modified through the SBAPI (specifically via SetLine).
When this flag is set, IsValid() no longer requires a valid address range,
only a valid line number.

This is risk-free because the flag is only set for LineEntry objects created
through the SBAPI, which are primarily used by scripted processes and frames.
Regular debug information-derived line entries continue to require valid
address ranges.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere commented:
Can we break this up into an NFC patch that makes `LineEntry` a class and a separate PR for the functional change that makes `AddressRange` optional?

medismailben commented:
> > Can we break this up into an NFC patch that makes `LineEntry` a class and a separate PR for the functional change that makes `AddressRange` optional?
> 
> +1.
> 
> Why do you want to turn it into a class? In C++, the only difference between a class and a struct is the default access level. Structs are public by default, Classes are private by default.

To Alex's point, this didn't require to turn the `LineEntry` struct in to a class so I just made the `range` member a private optional and added public getter/setter methods.

JDevlieghere reviewed (changes requested):
I strongly dislike this API because it easy to misuse and hard to use correctly. That's based on my very recent experience with the SupportFileSP. When it's used incorrectly (which it will, just like SupportFileSP was) it will lead to subtle and hard to diagnose bugs. My main qualm is the `GetRange` accessor which gives a false sense of security. If we need to make the `AddressRange` optional, then we should just keep it a (public) member and force every call site to consider that it may not be present.

bulbazord reviewed (changes requested):
Instead of modifying the AddressRange member to be an optional, why not *only* relax the `LineEntry` validity check? There's a lot of code that needs to be changed here to make the `range` member optional and you need to remember to call `HasValidRange`. Not every place that accesses the AddressRange checks for a valid range either.

Functionally, the only benefit to having AddressRange be an optional is you can tell the difference between "The AddressRange was set to an invalid value" versus "The AddressRange member was untouched". Is there value in knowing the difference?

Requesting changes because my current impression is that this adds unnecessary complexity to LineEntry.

JDevlieghere reviewed (changes requested):
Can you please update the PR description with more details about why this is necessary? The current description doesn't explain the problem this is solving nor how this is the (best) solution. 

In the context of the line table it makes sense to require a line entry to have an address (range). This PR also needs to motivate why dropping this requirement makes sense. 

Finally, we have a bunch of places across LLDB that call `IsValid`. I see it referenced in `SymbolContext`, `StackFrame`, `Disassembler`, `FormatEntity`, `CommandObjectSource` to name a few. Did you look through all its uses and confirmed that this change in behavior remains correct? The fact that the test suite passes is promising but far from a guarantee as I doubt someone went out of their way to cover this previously unexpected situation. 

I'm not saying this isn't the right approach, but right now there's not enough information in this PR to come to the conclusion that it is.

clayborg reviewed (commented):
How does anyone find a `LineEntry` with no address ranges. Is there anything we can put into the address fields of a `LineEntry` that has no PC value that helps to identify where it comes from? Code or byte offset? Is this for a specific target like wasm?

clayborg commented:
Lots of things might break in LLDB if the `LineEntry::IsValid()` changes. `LineEntry` objects are used for single stepping as those rely on address ranges for doing this stepping and if a `LineEntry` claims it is valid, we might try to single step with no valid address range. I worry that this change might break a lot of things in LLDB.

clayborg commented on lldb/source/API/SBLineEntry.cpp:
```diff
@@ -60,7 +60,7 @@ SBAddress SBLineEntry::GetEndAddress() const {
+  if (m_opaque_up && m_opaque_up->range.IsValid()) {
     sb_address.SetAddress(m_opaque_up->range.GetBaseAddress());
     sb_address.OffsetAddress(m_opaque_up->range.GetByteSize());
   }
```
Might be the same issue as above where the resulting `sb_address` is the same as before this change unless the byte size is set to something valid, but the base address isn't. If the base address is invalid, it will be set to LLDB_INVALID_ADDRESS which is UINT64_MAX. So even if the byte size is set to something valid, it will result in us returning an address that was offset from UINT64_MAX, which won't be useful

medismailben replied:
ditto

bulbazord reviewed (approved):
I do like the `is_synthetic` member better than an optional `AddressRange` because it conveys the intent better. In the future, it might also be able to explain why another field might appear invalid or otherwise strange.

medismailben commented:
@clayborg Are you fine with this approach ?

clayborg commented on lldb/include/lldb/Symbol/LineEntry.h:
```diff
@@ -136,6 +136,10 @@ struct LineEntry {
 
+  /// This gets set for LineEntries created without a valid address range.
+  /// When set, `LineEntry::IsValid` doesn't check the `range` validity.
+  bool synthetic;
```
initialize this with false here just in case new constructors are added later?
```
bool synthetic = false;
```

clayborg reviewed (approved):
Looks good! Just a few inline comments that can be implemented or ignored.

medismailben commented on lldb/source/API/SBLineEntry.cpp:
```diff
@@ -132,6 +132,8 @@ void SBLineEntry::SetLine(uint32_t line) {
 
   ref().line = line;
+  if (!ref().range.IsValid())
+    ref().synthetic = true;
```
Right, we don't that's why I didn't bother making a setter for it for now.

medismailben committed:
[lldb] Add support for synthetic LineEntry objects without valid address ranges

Scripted Frames that materialize Python functions are PC-less by design,
meaning they don't have valid address ranges. Previously, LineEntry::IsValid()
required both a valid address range and a line number, preventing scripted
frames from creating valid line entries for these synthetic stack frames.

Relaxing this requirement is necessary since `SBSymbolContext::SetLineEntry`
will first check if the LineEntry is valid and discard it otherwise.

This change introduces an `synthetic` flag that gets set when LineEntry
objects are created or modified through the SBAPI (specifically via SetLine).
When this flag is set, IsValid() no longer requires a valid address range,
only a valid line number.

This is risk-free because the flag is only set for LineEntry objects created
through the SBAPI, which are primarily used by scripted processes and frames.
Regular debug information-derived line entries continue to require valid
address ranges.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
